1D Wavelength Calibration

hpparvi opened the pull request:
This PR introduces a new **1D wavelength calibration** class, `specreduce.wavecal1d.WavelengthCalibration1D`, intended to fully replace the existing `specreduce.wavelength_calibration.WavelengthCalibration1D` in version 2.0 (I'm keeping the previous class for backwards compatibility).

The only (somewhat) significant feature currently missing is support for generating standard spectroscopic `-GRI` and `-GRA` WCSs as described in [Greisen et al. (2006)](https://ui.adsabs.harvard.edu/abs/2006A%26A...446..747G/abstract). At the moment, the class supports only GWCS-based WCS. This is something we can address in a future update.

hpparvi committed:
Added a new approach to 1D wavelength calibration based on minimizing a sum of nearest-line distances between a theoretical line list and a transformed observed line list.

hpparvi committed:
Fixed a minor 1D wavelength solution rebinning issue.

hpparvi committed:
Added utility functions to differentiate 1d and 2d polynomial models with respect to x.

hpparvi committed:
Renamed and refactored the 1D wavelength calibration module.

Renamed `wavecal.py` to `lswavecal1d.py` and refactored the `WavelengthSolution1D` class for improved clarity and functionality. Added new methods like `refine_fit`, `pix_to_wav`, and `wav_to_pix`, and modified existing ones to enhance flexibility and maintainability.

hpparvi committed:
Added 2D wavelength calibration class and associated methods.

Introduce `WavelengthSolution2D` class for 2D wavelength calibration, extending functionality from 1D calibration.

hpparvi committed:
Introduced a 2D WCS property that constructs a WCS pipeline using pixel and spectral frames.

hpparvi committed:
Started working on 1D wavelength calibration documentation.

hpparvi committed:
Cleaning up and documenting the 1D wavelength calibration code.

hpparvi committed:
Continuing 1D spectrum calibration documentation and cleanup.

hpparvi committed:
Refactored and enhanced WavelengthSolution1D class methods to improve type hints, documentation, and method structure. Added new methods for calculating inverse transformations and derivatives. Enhanced plotting functionality with additional parameters and refined docstrings for clarity.

hpparvi committed:
Added unit handling for wavelength and RMS/residual plotting.

hpparvi committed:
Added a basic example notebook for 1D wavelength calibration with Specreduce.

hpparvi committed:
Renamed some parameters for clarity in 2D wavelength calibration.

hpparvi committed:
Updated the 1D wavelength calibration plotting routines handle cases where `arc_spectra` is `None`.

hpparvi committed:
Updated logic to handle cases where `arc_spectra` is `None` and added `plim` parameter for pixel range control in plotting.

hpparvi committed:
Refactored `WavelengthSolution1D` to improve readability, flexibility, and modularity, while resolving associated issues. Enhanced methods for handling pixel-to-wavelength mappings, line-matching processes, and plotting capabilities. Added new functions and properties for better control over data and fit refinement.

hpparvi committed:
Replaced internal `_lines_pix` and `_lines_wav` attribute modifications with setters to ensure proper masking and validation. Added `plot_lines_pix` and `plot_lines_wav` methods for enhanced modularity in plotting observed and theoretical lines.

hpparvi committed:
Fixed line matching to work with masked arrays.

hpparvi committed:
Fixed transformation to work with masked arrays.

hpparvi committed:
Updated the handling of observed and catalog lines. Introduced new methods, updated property names, and enhanced plotting functionality.

hpparvi committed:
Introduced new optional parameters like `refine_fit` and enhanced documentation for plotting and fitting methods.

hpparvi committed:
Renamed `lines_pix` to `observed_lines` and `lines_wav` to `catalog_lines`.

hpparvi committed:
The refine_fit method now supports a max_iter parameter to control the maximum number of iterations for refining the pixel-to-wavelength transformation. This ensures the process adjusts until convergence or the specified iteration limit is reached.

hpparvi committed:
Added 1D wavelength calibration tests.

hpparvi committed:
Updated WaveCal1D tests.

hpparvi committed:
Fixed minor issues while writing tests.

hpparvi committed:
Updated tests.

hpparvi committed:
Added tests for wavelength-pixel transformations and WCS creation.

hpparvi committed:
Added new tests for WavelengthSolution1D and plotting methods.

hpparvi committed:
Removed redundant warnings and improved error handling.

hpparvi committed:
Polished the tests.

hpparvi committed:
Cleaned up an error message.

hpparvi committed:
Removed 2D wavelength calibration code from this branch. Will work on it on another one. Also removed the old 1D wavelength calibration code.

hpparvi committed:
Renamed WavelengthSolution1D -> WavelengthCalibration1D.

hpparvi committed:
Added the first version of the WavelengthCalibration1D docs with Jupyter Notebook tutorials.

hpparvi committed:
removed a reference to old wavelength_calibration module from the main __init__.py

hpparvi committed:
Updated the wavelength calibration tutorial notebooks.

hpparvi committed:
Introduced the `value_fontsize` parameter to control font size in spectral line plots across various methods and adjusted vline colors. Fixed minor formatting issues for consistency.

hpparvi committed:
Revised the wavelength calibration notebooks.

hpparvi committed:
Finalised 1D wavelength calibration tutorial 1.

hpparvi committed:
Changed the wcs attribute to gwcs.

hpparvi committed:
Finalised 1D wavelength calibration tutorial 2 and polished the first tutorial a bit.

hpparvi committed:
Finished 1D wavelength calibration tutorial 3.

hpparvi committed:
Cleaned up the wavelength calibration docs.

hpparvi committed:
Codestyle fixes for the 1D wavelength calibration code.

hpparvi committed:
Changed the documentation API.

hpparvi committed:
Brought back the original wavelength_calibration.py code so that we don't break backwards compatibility (quite yet).

hpparvi committed:
Fixed a broken WavelengthCalibration1D tests.

hpparvi committed:
Added a docstring for the WavelengthCalibration1D initialiser.

hpparvi commented:
Ok, looks like all the builds are failing because of a matplotlib dependency that I added. I'll need to make matplotlib and the plotting functionality optional.

hpparvi committed:
Small wavelength calibration documentation fixes.

hpparvi commented:
Docs build works with Jupyter Notebook examples. Take a look at https://specreduce--265.org.readthedocs.build/en/265/wavelength_calibration/wavelength_calibration.html

hpparvi committed:
Updated the changelog.

hpparvi committed:
Made sure the basic WavelengthCalibration1D functionality can be used without matplotlib installed.

hpparvi committed:
Made matplotlib a basic dependecy.

hpparvi committed:
Revised wavelength calibration tests.

hpparvi committed:
Codestyle fix.

hpparvi committed:
Added a TiltCorrection class for 2D rectification.

hpparvi committed:
Refactored the tilt correction code to improve readability, structure, and functionality. Added detailed docstrings, refined methods for arc line identification, transformation fitting, and matching lines. Improved internal handling of parameters and optimized logic.

hpparvi committed:
Added type annotations and a docstring for the `diff_poly2d_x` function.

hpparvi committed:
Renamed private methods to public and adjusted their parameters for consistency and improved usability. Updated variable names and docstrings for better clarity and alignment with the code's functionality. Simplified return behavior in `match_lines` to support optional concatenation.

hpparvi committed:
Introduced a method to transform coordinates from rectified space to detector space.

hpparvi committed:
Refactored plot functionality and add fit quality visualization.

hpparvi committed:
Added tilt correction tutorial and sample data.

hpparvi committed:
Introduced `disp_axis` and `mask_treatment` parameters to enhance flexibility in 2D spectrum rectification. Updated methods to use these parameters, ensuring proper alignment and control over mask handling. Also improved documentation and refactored variable naming.

hpparvi committed:
Added basic unit tests for TiltCorrection. Cover that the procedure runs, but must be improved to ensure that the results are correct.

hpparvi committed:
Updated docs related to tilt correction.

hpparvi committed:
Codestyle fixes.

hpparvi committed:
Documentation fixes.

hpparvi committed:
Documentation hack. Needs to be addressed correctly.

cshanahan1 commented on notebook_sandbox/wavecal_1d_shane.ipynb:
```diff
@@ -0,0 +1,283 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "flux = getdata('shane_kast_blue_600_4310_d55.fits', 1).astype('d')\n",
```
need to provide the path `../docs/wavelength_calibration/wavelength_calibration/`, and also it looks like the file in that directory is different shane_kast_blue_600_4310_d55.fits so either change the filename to that one or add the correct file.

cshanahan1 commented on notebook_sandbox/wavecal_1d_shane.ipynb:
```diff
@@ -0,0 +1,283 @@
+   ],
+   "source": [
+    "%%time\n",
+    "ws = WavelengthSolution1D(arc_spectra=arc_spectrum, line_lists=[['CdI', 'HgI', 'HeI']],  wlbounds=(3200, 5700))\n",
```
This class doesn't exist, if this should replaced with WavelengthCalibration1D and wlbounds needs to be line_list_bounds and ref_pixel should be supplied. Also remove %%time from the top of this cell. I got a traceback when I tried this replacement with 'ref_pix', so i'll wait to continue reviewing the rest of this notebook

cshanahan1 reviewed (commented):
I gave this a first pass and left a few comments. Notebook outputs need to be cleared, and i think it might make sense to have a data subdirectory rather than putting these files in docs

hpparvi committed:
Removed an old Jupyter Notebook file.

hpparvi committed:
Added `matplotlib>=3.7` as a testing dependency in `pyproject.toml`.

hpparvi commented on notebook_sandbox/wavecal_1d_shane.ipynb:
```diff
@@ -0,0 +1,283 @@
+    "from matplotlib.pyplot import setp, subplots, close, rc\n",
+    "from specutils import Spectrum1D\n",
+    "\n",
+    "from specreduce.lswavecal1d import WavelengthSolution1D\n",
```
Thanks for catching this @cshanahan1. This notebook was outdated, and I shouldn't have included it in the PR in the first place.

hpparvi committed:
Replaced `Spectrum1D` with `Spectrum` throughout `wavecal1d.py` for compatibility updates.

hpparvi committed:
Revised wavecal1d tests.

hpparvi committed:
Fixed `flux` assignment in `tilt_correction.py` to ensure proper handling of unit-aware array values.

hpparvi committed:
Refactored observed line handling in `wavecal1d.py`, added amplitude support, and improved the plotting of observed lines.

hpparvi committed:
Improved uncertainty type handling in WavelengthCalibration1D.

hpparvi committed:
Added "Spectrum" to ``__all__` in ``specreduce.compat``.

tepickering commented on docs/tilt_correction/common.py:
```diff
@@ -0,0 +1,19 @@
+from pathlib import Path
+
+from astropy import units as u
+from astropy.io import fits as pf
```
why is `fits` getting aliased to `pf` here? seems a bit confusing...

hpparvi replied:
Whoops, this is a leftover habit from PyFITS times (import pyfits as pf). I'll fix it in the TiltCorrection PR.

tepickering commented:
gave it another read-through and it's looking good! very ambitious amount of work done here. looks like the tests are not happy so once they're sorted out, i'll do a deeper dive and try things out for myself locally.

hpparvi committed:
Refined handling of arc line amplitudes in 1D wavelength calibration. Updated logic to compute amplitudes directly from flux near centroids for improved plotting reliability.

hpparvi committed:
Refactored handling of observed_lines and catalog_lines properties.

hpparvi committed:
Enhanced line label placement and styling in 1D wavelength calibration plots. Added text unclutter functionality and updated subplot axes to share x and y scales.

hpparvi committed:
Updated the wavecal1d tests.

hpparvi committed:
Codestyle fixes.

hpparvi committed:
Removed the tilt-reduction related code and docs.

hpparvi committed:
Revised docs.

hpparvi committed:
Refactored internal handling of line lists. Introduced `_format_linelist` utility for standardizing input data. Added cached properties for observed and catalog line locations and amplitudes. Enhanced line-matching logic to improve code structure and reliability. Simplified plotting functions and improved line label rendering.

hpparvi committed:
Codestyle fixes.

hpparvi committed:
Revised 1D wavelength calibration tests.

hpparvi committed:
Removed wavelength calibration tutorial files and data.

hpparvi committed:
Refactored `plot_labels` parameter in 1D wavelength calibration plots to support sequence inputs. Introduced `_create_trees` helper method for managing KDTree objects. Improved line label rendering and axis scaling. Updated tests to reflect these changes.

hpparvi committed:
Fixed typos and improved clarity in `wavecal1d` docstrings and comments. Renamed `remove_ummatched_lines` to `remove_unmatched_lines`. Updated corresponding test.

hpparvi committed:
Refactored `wavecal1d` module: improved docstrings and comments for clarity, added type hints, enhanced error handling, and optimized label uncluttering logic.

hpparvi committed:
Changed to allow WavelenghtCalibration1D initialisation without an observed line list or spectra.

hpparvi committed:
Revised documentation.

hpparvi committed:
Added return type hint `-> None` for `remove_unmatched_lines` method in `wavecal1d`.

hpparvi committed:
Fixed docstring formatting in `wavecal1d`.

hpparvi committed:
Improved label rendering precision in 1D wavelength calibration plots by rounding to significant figures dynamically.

hpparvi committed:
Sort the catalog line lists by wavelength.

hpparvi committed:
Handled NaN values in flux operations using `np.nansum` and `np.nanmax` in `wavecal1d`.

hpparvi committed:
Revised the wavelength calibration docs.

hpparvi committed:
Enhanced error handling and optimization in `wavecal1d`: added validation for `degree` and `ref_pixel`, introduced `higher_order_limits` for coefficient bounds, and refined differential evolution bounds calculation.

hpparvi committed:
Updated `matplotlib` dependency in `pyproject.toml` to `>=3.10`.

hpparvi commented:
Thanks @tepickering! I've removed the 2D tilt correction code and tutorials from this PR, and I have continued revising the code and docs based on new datasets I've tested the code with. Feel free to take a deeper look.

The only thing I may still want to change is the name of the `fit_global` method. This method is intended for non-interactive workflows (automatic pipelines with relatively stable instruments), while the `fit_lines` is meant for interactive workflows. The main difference between the two is that `fit_lines` requires a set of user-provided matched lines in pixel and wavelength space, and `fit_global` doesn't.  The "global" in the name refers to the method's use of global optimization for calculating the wavelength solution. However, the name doesn't describe what the method does or explain how it differs from `fit_lines`. Does anyone have good alternative names for the method?

kbwestfall reviewed (commented):
Not finished yet, but I wanted to get these comments out ahead of the meeting tomorrow.  I'll try to finish tomorrow morning my time.

kbwestfall reviewed (commented):
This is the last batch of comments for now.  I didn't look closely at the plotting or testing functions.  Happy to talk through any of these (from this batch or the previous one).  Thanks for all the hard work on this!!

hpparvi committed:
Replaced `gwcs.coordinate_frames` alias with full module path for clarity in `wavecal1d`.

hpparvi committed:
Revised the code to use np.ma.getmaskarray to get a MaskedArray mask as an array.

hpparvi committed:
Improved wavecal1d.py type hints.

hpparvi commented on docs/wavelength_calibration/wavelength_calibration.rst:
```diff
@@ -0,0 +1,247 @@
+calibration** is the process of determining this relationship, creating a mapping function or
+model  that converts pixel coordinates to wavelength values.
+
+This is often achieved by observing a calibration source with well-known emission or absorption
```
Iodine absorption cells are used to calibrate high-resolution radial velocity measurements in exoplanet research (for example, [Marcy and Butler, 1992](https://articles.adsabs.harvard.edu/pdf/1992PASP..104..270M)), but this is quite a different thing from what we're doing here, so I've removed the mention about absorption lines.

hpparvi commented on docs/wavelength_calibration/wavelength_calibration.rst:
```diff
@@ -0,0 +1,247 @@
+-------------------------
+
+The `~specreduce.wavecal1d.WavelengthCalibration1D` class encapsulates the data and methods
+needed to perform  1D wavelength calibration. The class supports multiple workflows with varying
```
True, there's quite a bit of unnecessary filler text here. It parallels the tutorials that demonstrate how to use the class in both interactive and non-interactive workflows, but this level of verbosity isn't really necessary in documentation.

hpparvi commented on docs/wavelength_calibration/wavelength_calibration.rst:
```diff
@@ -0,0 +1,247 @@
+*   **Using an Arc Spectrum**: Provide the arc spectrum as a `specutils.Spectrum`
+    object via the ``arc_spectra`` argument. You also need to provide a ``line_lists`` argument,
+    which can be a list of known catalog wavelengths or the name(s) of standard line lists
+    recognized by `specreduce` (e.g., ``"HeI"``).
```
The available line lists can be accessed via `specreduce.calibration_data.get_available_line_catalogs`. I have now added a mention of this to the docs.

tepickering commented on pyproject.toml:
```diff
@@ -20,16 +20,19 @@ dependencies = [
 ]
 all = [
-    "matplotlib>=3.7",
+    "matplotlib>=3.10",
```
agreed.

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+            raise ValueError("Must provide arc spectra to find lines.")
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
```
Good catch! This was a leftover from playing around with things.

hpparvi commented on docs/wavelength_calibration/wavelength_calibration.rst:
```diff
@@ -0,0 +1,247 @@
+
+    .. code-block:: python
+
+        wc = WavelengthCalibration1D(ref_pixel=512,
```
True, `ref_pixel` is not a keyword argument, but I prefer to write argument names explicitly in these examples because it makes them easier to understand. That is, it is directly clear that the first argument that we're giving is the `ref_pixel`.

We can make `ref_pixel` optional and set it to default to the centre pixel of the provided spectrum, but this only works when the class is instantiated with a spectrum (or spectra). I'll modify the code so that `ref_pixel` is required only when spectra are not supplied at instantiation.

hpparvi commented on docs/wavelength_calibration/wavelength_calibration.rst:
```diff
@@ -0,0 +1,247 @@
+    print(wc.observed_line_locations)
+
+This populates the `~specreduce.wavecal1d.WavelengthCalibration1D.observed_line_locations`
+attribute.
```
Fixed.

hpparvi commented on docs/wavelength_calibration/wavelength_calibration.rst:
```diff
@@ -0,0 +1,247 @@
+    When ``refine_fit=True`` is set, the method automatically identifies matching pairs between
+    observed and catalog lines, then performs a least-squares refinement using **all matching lines**.
+    This goes beyond the subset of lines provided to :meth:`~specreduce.wavecal1d.WavelengthCalibration1D.fit_lines`,
+    resulting in a more complete wavelength calibration.
```
Sure, `fit_lines` fits a model to a group of user-identified lines. This can generally be a small subset of the actual observed lines, but will generate a good initial transformation. `refine_fit` uses the existing model to match all the observed lines with the catalogue lines, and then refines the model by fitting all the matched lines (most of, or all the observed lines if the initial model was sufficiently accurate). This way, the final model will be calculated using as many lines as possible, even if the user only gives a small number of lines for the `fit_lines` method.

hpparvi committed:
Revised wavelength calibration documentation.

hpparvi committed:
Made matplotlib a primary dependency.

hpparvi committed:
Typing cleanup.

hpparvi committed:
- Made `ref_pixel` an optional argument in `WavelengthCaliobration1D`.
- Stopped catching warnings in `WavelengthCaliobration1D.find_lines`.
- Revised the code style.

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+        self.catalog_lines = [lst.compressed().reshape([-1, 2]) for lst in self._cat_lines]
+        self._create_trees()
+
+    def rms(self, space: Literal["pixel", "wavelength"] = "wavelength") -> float:
```
We could, but even then, I'd prefer to calculate the RMSs using a function. And if we have the function available, it makes little difference to access the RMS by calling the function rather than via an attribute or ;) So, in my opinion, a method is the cleanest way to access the RMSs. Do you have any specific use case in mind where saving them as attributes would be better?

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+                rms = rms_new
+        self._calculate_p2w_derivative()
+        self._calculate_p2w_inverse()
+        self.match_lines(max_match_distance)
```
No really, fixed.

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+                bounds.append(np.asarray([v, v]))
+        else:
+            for i in range(2, model[1].degree + 1):
+                bounds.append(np.array([-1, 1]) * 10**(np.log10(np.mean(dispersion_bounds)) - 2))
```
I wrote these lines while flying and left a small bug in the bounds initialization. The code should be:

```Python
for i in range(2, model[1].degree + 1):
    bounds.append(
        np.array([-1, 1]) * 10 ** (np.log10(np.mean(dispersion_bounds)) - 2*i)
)
```

That is, the second-order bounds are 1e-4 of the first-order (dispersion) bounds, and each higher order is tightened by two additional orders of magnitude. I don’t have a strong theoretical justification for this schedule,
but this approach worked best with the different spectra I was testing the code against. In most cases, the user will want to fine-tune the higher-order limits for the instrument setup to speed up the convergence.

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+            for t, l in zip(self._trees, obs_lines):
+                transformed_lines = model.evaluate(l, -self.ref_pixel, *x)[:, None]
+                total_distance += np.clip(t.query(transformed_lines)[0], 0, max_distance).sum()
+            return total_distance
```
I find locally defined functions like this to be useful, especially to reduce clutter, and this was my main motivation for defining the function inside the optimization method. Then again, it can be useful to have access to our cost function to test different optimizers, etc., so I've moved the function as a private `WavelengthCalibration1D` method.

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+
+        refine_max_distance
+            The maximum allowed distance between the catalog and observed lines for them to be
+            considered a match.
```
I've revised the docstring to mention that this argument is ignored when `refine_fit=False`, and added `refine_max_distance` to `fit_dispersion` (was `fit_global`).

hpparvi committed:
Revised `fit_lines`.

hpparvi committed:
Refactored automatic wavelength calibration:
 - Added `_line_match_distance` method
 - Renamed `fit_global` to `fit_dispersion`
 - Fixed bugs in setting the default higher order coefficient boundaries

hpparvi committed:
Improved wavelength calibration loop by using `math.isclose` for RMS calculation.

hpparvi committed:
Switched `gwcs` property to use `cached_property` and removed unused `_wcs` attribute in `wavecal1d.py`.

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+                raise ValueError("Cannot fit without catalog lines set.")
+            tree = KDTree(
+                np.concatenate([c.compressed() for c in self.catalog_line_locations])[:, None]
+            )
```
Exactly, these temporary trees combine all the line lists. Here, our goal is to match the provided approximate line locations (in pixel and wavelength) to the full-precision lines. We don't need to distinguish between the different arc lamps (we could, but I don't think this would be useful). Instead, we just need to identify the nearest full-precision lines from the combined line lists corresponding to the given ones.

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+        Optionally, the provided pixel and wavelength values can be "snapped"
+        to the nearest values present in the internally stored observed line
+        list and catalog line list, respectively. This can correct for small
+        inaccuracies in the input pairs if the internal lists are populated.
```
The motivation behind the matching functionality is to make life easier for lazy users who don't want to input the lines with full precision ;) When matching is enabled (and the internal line lists exist), it is sufficient to input the lines with a precision that allows them to be matched with the stored full-precision lines.

hpparvi replied:
The reason I wanted to keep `line_lists` optional is that we can instantiate an "empty" `WavelengthCalibration1D` object. This is useful so that we can save and load wavelength solutions easily as GWCS (not yet implemented, but should be done in the near future).

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+                # the centroids would be fine. The amplitudes are used only for plotting,
+                # so selecting the maximum flux near the line centroid is the best approach.
+                amplitudes = np.zeros(len(lines))
+                for j, line in enumerate(lines["centroid"].value):
```
This works beautifully, thanks! I think the call to `maximum_filter1d` is fine. A single call typically takes only a few microseconds (for an array of ~2000 elements).

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+    A new Polynomial1D model representing the derivative of the input Polynomial1D model.
+    """
+    coeffs = {f"c{i-1}": i * getattr(m, f"c{i}").value for i in range(1, m.degree + 1)}
+    return models.Polynomial1D(m.degree - 1, **coeffs)
```
I can take a look into this. However, getting this accepted to astropy core will take time, so I'd prefer to have it here as a helper function for now.

hpparvi committed:
Improved amplitude estimation for found lines.

hpparvi committed:
Allow for `line_list_bounds` to be `None`.

hpparvi committed:
Minor docstring and code style fixes.

hpparvi commented on specreduce/compat.py:
```diff
@@ -1,7 +1,7 @@
 from astropy.utils import minversion
 
-__all__ = []
+__all__ = ["Spectrum"]
```
I'm not convinced we should do this quite yet. The changes in this PR don't break our backwards compatibility, and we have a working compatibility layer with `specutils < 2`. If we change the dependency to require `specutils > 2`, we break all code that uses `specreduce` and `specutils` that isn't yet updated to work with `specutils >= 2`. 

This type of code may not be common, but it would still be inconvenient to some users. So, I'd rather change the requirement later.

hpparvi committed:
Made `line_list_bounds` default to `None` in the initializer.

hpparvi committed:
Refactored wavelength calibration to support dynamic polynomial degree updates:
- Introduced `degree` as an adjustable parameter in `fit_lines`, `refine_fit`, and `fit_dispersion`.
- Removed fixed `degree` from the constructor, now initialized dynamically.
- Updated `_init_model` to accept optional coefficients.
- Improved handling of refined fits and degree validation via a `degree` property.

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+        pix_bounds: tuple[int, int] | None = None,
+        line_list_bounds: tuple[float, float] = (0, np.inf),
+        wave_air: bool = False,
+    ) -> None:
```
I've now made `ref_pixel` optional when `arc_spectra` is provided, in which case `ref_pixel` defaults to the centre of the spectrum. 

I've now refactored the `degree` handling. I've removed the `degree` argument from the constructor and added it to the two fitting methods. I've also tidied up the model initialisation a bit and included a `degree` argument in `refine_fit`. If the `refine_fit` degree exceeds the current model degree, `refine_fit` will compute a higher-degree model based on the existing coefficients. This was how I originally designed `refine_fit` to function, but I changed it at some point along the way. 

I believe `line_list_bounds` should remain in the constructor because it will actually alter the loaded line lists, and the approximate wavelength range covered by the observation setup should still be quite well known in advance.

hpparvi committed:
Revised the `resample` method. The previous version had several errors what comes to flux conservation and uncertainty handling. These should all be fixed now.

hpparvi committed:
Updated wavelength calibration tests.

hpparvi committed:
Wavelength calibration cleanup.

hpparvi committed:
Revised wavelength calibration documentation.

hpparvi committed:
Continued revising wavelength calibration documentation.

hpparvi committed:
Revised wavelength calibration tests.

hpparvi commented on specreduce/wavecal1d.py:
```diff
@@ -0,0 +1,1240 @@
+        fitter = fitting.LinearLSQFitter()
+        rms = np.nan
+        for i in range(max_iter):
+            self.match_lines(max_match_distance)
```
That’s a very good point I haven’t run into problems with this in practice, but it’s a theoretical risk worth keeping in mind. It will need some testing and thought, so I’ve opened issue #288 to track it for later.

hpparvi commented:
Thanks for your work, @kbwestfall. This was an excellent, thorough, and useful review! I've now (hopefully) addressed all your comments, so please feel free to take a look.

hpparvi committed:
Refactored `wavecal1d` to use a new `WavelengthSolution1D` class for managing wavelength models and transformations.

hpparvi committed:
Refactored wavelength solution handling:
- Introduced `WavelengthSolution1D` to manage pixel-to-wavelength transformations.
- Updated references to `_solution` and `p2w` for clarity and consistency.
- Added new tests for `WavelengthSolution1D`.

hpparvi committed:
Improved WavelengthSolution1D tests.

hpparvi committed:
Expanded and updated `WavelengthSolution1D` docstring and improved property type annotations.

hpparvi committed:
Removed `ref_pixel` attribute from `WavelengthSolution1D` and adjusted related tests and docstrings. Expanded and clarified docstring for `WavelengthSolution1D` initialization.

hpparvi committed:
Refactored `WavelengthCalibration1D` and `WavelengthSolution1D`:
- Replaced `_solution` with `solution` for clarity.
- Updated return types of fit methods to return `WavelengthSolution1D`.
- Reorganized parameter order and docstrings for consistency.

hpparvi committed:
Updated wavelength calibration tests.

hpparvi committed:
Updated wavelength calibration documentation to include `WavelengthSolution1D` usage and examples:
- Added `automodapi` entry for `specreduce.wavesol1d`.
- Expanded explanations of 1D calibration workflow with `WavelengthSolution1D`.
- Updated code examples to reflect new API.

hpparvi committed:
Revised wavelength calibration documentation.

hpparvi committed:
Linked `bounds_pix` to `solution.bounds_pix` in `WavelengthCalibration1D`.

hpparvi committed:
Added `n_strongest_lines` parameter to filter the strongest spectral lines for wavelength calibration

cshanahan1 commented on specreduce/wavesol1d.py:
```diff
@@ -0,0 +1,246 @@
+            1D spectrum binned to the specified wavelength bins.
+        """
+        if nbins is not None and nbins < 0:
+            raise ValueError("Number of bins must be positive.")
```
and also non-zero?

hpparvi replied:
Good point. I'll fix this.

cshanahan1 reviewed (commented):
No major comments, this looks great!

I saw in some of your initial commits to this that you had some jupyter notebooks to demonstrate use, I think it would be nice to have those back if they're in decent shape.
